### PR TITLE
SAM pattern

### PR DESCRIPTION
**Pull request by vladislavsimic**

### CODE REVIEW

**Comment by npathai** [comment]
@iluwatar I will review it.

**Comment by npathai** [comment]
@vladislavsimic Hey. Firstly thank you for contributing. I will complete review by tomorrow. Is that okay with you?

**Comment by vladislavsimic** [comment]
@npathai Ok.

File: sam/pom.xml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <parent>
+    <artifactId>java-design-patterns</artifactId>
+    <groupId>com.iluwatar</groupId>
+    <version>1.16.0-SNAPSHOT</version>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+  
+  <artifactId>sam</artifactId>
```
Comment (npathai) [inline sam/pom.xml]: Add sam to parent pom.

File: sam/src/main/java/com/iluwatar/sam/StateEngine.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    this.actions = actions;
+  }
+
+  /**
+   * Method for render model 
+   * 
+   * @param model from RocketModel
+   */
+  public void render(RocketModel model) {
+    if (model.getCounter() == 0) {
```
Comment (npathai) [inline sam/src/main/java/com/iluwatar/sam/StateEngine.java]: Doubt: Half of our state i.e. Counter is inside Model and other half i.e. current state of rocket launcher is represented in terms of `RocketState` implementation. I think for current example it works, but how can we add abort action to this?
As soon as we introduce abort action, then we also introduce additional semantics that a RocketLauncher that is in Ready state / Launched state cannot be aborted. Also hypothetically if we get a start action after Abort we need to factor in the time required for technical team to check launchpad (say for 20 secs) and then again start the counter.

File: sam/src/test/java/com/iluwatar/sam/SamPatternTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    model = new RocketModel();
+    actions = new RocketActions(model);
+    view = new RocketView(actions);
+    stateEngine = new StateEngine(view, actions);
+    
+  }
+  
+  @Test
+  public void testPattern() throws Exception {
+    assertNotNull(model);
```
Comment (npathai) [inline sam/src/test/java/com/iluwatar/sam/SamPatternTest.java]: Assertion checks for members that you have initialized in setup are not required.

**Comment by npathai** [comment]
@vladislavsimic You have my initial review. Well I must say good initial implementation. I am not a front end developer so some suggestions may not apply. But for someone like me (who is a noob) in this area, just extending implementation further, abort action, will help a lot in understanding CRCs (Class Responsibility and Collaborators). Thanks.

**Comment by vladislavsimic** [comment]
Hi @npathai. Can I implement the view in the java swing library or the like? The other option is to start an abort action by placing a flag on which the counter will stop?

**Comment by npathai** [comment]
@vladislavsimic Sure you can implement it using Swing. Or else a very basic javascript web app (just buttons and countdown header) will do as well. Its up to your personal preference.

**Comment by npathai** [comment]
@vladislavsimic Any updates? Should I review the refactored changes?

**Comment by abhayamuni** [comment]
can the administrator of this thread please unsubscribe me so that I no longer receive emails.  I have tried to remove myself for months now unsuccessfully (i.e., mute thread, disable via web interface, etc).

Please help,
Joseph

**Comment by iluwatar** [comment]
Pull request closed due to inactivity

**Pull request closed without merging**